docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+==================================
+RFC 010: Peer to Peer Light Client
+==================================
+
+Changelog
+---------
+
+- 2022-01-21: Initial draft (@tychoish)
+
+Abstract
+--------
+
+The dependency on access to the RPC system makes running or using the light
+client more complicated than it should be, because it is difficult to get
+access to the RPC endpoints on a running node. There is no deep dependency for
+the light client on the RPC system, and there is a persistent notion that
+"make a p2p light client" is a solution to this operational limitation. This
+document explores the implications and requirements of implementing a
+p2p-based light client, as well as the possibilities afforded by this
+implementation.
+
+Background
+----------
+
+High Level Design
+~~~~~~~~~~~~~~~~~
+
+From a high level, the light client P2P implementation, is relatively straight
+forward, and the P2P-backed statesync implementation that took place during
+the 0.35 cycle. The light client only really needs to be able to request (and
+receive) a `LightBlock` at a given height. To support this, a new Reactor
+would run on every full node and validator which would be able to service
+these requests. The workload would be entirely request-response, and the
+implementation of the reactor would likely be very straight forward, and the
+implementation of the provider is similarly relatively simple. 
+
+The complexity of the project focuses around peer discovery, handling when
+peers disconnect from the light clients, and how to change the current P2P
+code to appropriately handle specialized nodes.
+
+I believe it's safe to assume that much of the current functionality of the
+current ``light`` mode would *not* need to be maintained: there is no need to
+proxy the RPC endpoints over the P2P layer and there may be no need to run a
+node/process for the p2p light client (e.g. all use of this will be as a
+client.) 
+
+The ability to run light clients using the RPC system will continue to be
+maintained.
+
+LibP2P
+~~~~~~
+
+While some aspects of the P2P light client implementation are orthogonal to
```

Comment:
This may also be a good testbed for the migration, since it is a use that doesn't currently exist in exactly this form.